[WIP] Extensions extraction for DTO enum fields

anthonycr opened the pull request:
# Summary
The nature of our API is such that enum values may be added that are publicly available and are also backwards compatible. An example is if a new account type is created, that account type must be accessible to old consumers of the API. As such, we cannot guarantee that the list of enums shipped with any specific version of the library will be comprehensive, and so we must make a fallback available.

## Implementation
I created an inline extension function on `String` that can convert the value to an `enum`. The enum must also implement a `StringValue` interface. Using this `asEnum` function, extensions can be created on a DTO that allow an enum value to be extracted from a `String` property on the DTO. It does this by grabbing all the enum values from the enum and checking the `StringValue.value` property that each enum is required to provide against the `String` provided. So for instance, on the `User` class, an extension property was created called `accountType: AccountType` and it converts the `account` property to its enum representation or defaults to `AccountType.UNKNOWN` if the type is unknown. This extension function can be consumed from Java or Kotlin:

```java
final User user = ...
final AccountType accountType = UserExtensions.getAccountType(user);
```

```kotlin
val user = ...
val accountType = user.accountType
```

As this is a WIP, here are the enums I converted so far are:
- AccountType
- AlbumLayoutType
- AlbumPrivacyViewValue
- AlbumThemeType
- ApiOptionsType

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/TranscodeType.kt:
```diff
@@ -1,17 +1,12 @@
 package com.vimeo.networking2.enums
 
-import com.squareup.moshi.Json
+enum class TranscodeType(override val value: String?) : StringValue {
```
This class is missing kdocs

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/UploadSpaceType.kt:
```diff
@@ -1,14 +1,10 @@
 package com.vimeo.networking2.enums
 
-import com.squareup.moshi.Json
+enum class UploadSpaceType(override val value: String?) : StringValue {
```
This class is missing kdocs

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/UploadStatusType.kt:
```diff
@@ -1,17 +1,12 @@
 package com.vimeo.networking2.enums
 
-import com.squareup.moshi.Json
+enum class UploadStatusType(override val value: String?) : StringValue {
```
This class is missing kdocs

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/UserActivityType.kt:
```diff
@@ -1,80 +1,55 @@
 
-    @Json(name = "add_comment")
-    ADD_COMMENT,
+    ADD_COMMENT("add_comment"),
```
This class is missing kdocs on the types

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/VideoActionType.kt:
```diff
@@ -1,23 +1,17 @@
 
-    @Json(name = "Added to")
-    ADDED_TO,
+    ADDED_TO("Added to"),
```
This class is missing kdocs on the types

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/VideoBadgeType.kt:
```diff
@@ -1,29 +1,21 @@
 
-    @Json(name = "cameo")
-    CAMEO,
+    CAMEO("cameo"),
```
This class is missing kdocs on the types

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/VideoQualityType.kt:
```diff
@@ -1,23 +1,17 @@
 
-    @Json(name = "hd")
-    HD,
+    HD("hd"),
```
This class is missing kdocs on the types

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/VideoSourceType.kt:
```diff
@@ -1,20 +1,14 @@
 package com.vimeo.networking2.enums
 
-import com.squareup.moshi.Json
+enum class VideoSourceType(override val value: String?) : StringValue {
```
This class is missing kdocs

howardrigberg commented on models/src/main/java/com/vimeo/networking2/enums/VideoStatusType.kt:
```diff
@@ -1,38 +1,27 @@
 
-    @Json(name = "available")
-    AVAILABLE,
+    AVAILABLE("available"),
```
This class is missing kdocs on the types

anthonycr commented on models/src/main/java/com/vimeo/networking2/Video.kt:
```diff
@@ -185,10 +189,11 @@ data class Video(
     @Json(name = "status")
-    @Deprecated("This field is deprecated in favor of upload and transcode.")
-    val status: VideoStatusType? = null,
+    @Deprecated("This property is deprecated in favor of upload and transcode.")
```
My view is that if a property is still available via the API, we should support it, even if it is deprecated. An additional reason to keep it is that we still heavily use the property in our Android application. 

Maybe @msya can weigh in on what he thinks.

msya replied:
For v1 of models, the goal is to have parity with our current models. I prefer keeping the status field for now as our app uses it. The API spec will in the future specify a deprecated property for fields and it will specify the version in which it was deprecated. This will help in annotating which fields should be deprecated. There is a ticket in the queue to handle API versioning and this problem would be addressed as part of that ticket.

anthonycr commented on models/src/main/java/com/vimeo/networking2/Album.kt:
```diff
@@ -139,3 +145,24 @@ data class Album(
+ * @see Album.layout
+ * @see AlbumLayoutType
+ */
+val Album.albumLayoutType: AlbumLayoutType
```
I don't have a strong preference either way. The code is definitely boilerplate and annoying to write, but it's not worryingly error prone or difficult to maintain as the `get() =` is straight forward and the underlying logic is shared between every extension. Discoverability _might_ suffer with the usage of annotation processing for those reading source code. However, the only concern I have is in relation to the reference in the property documentation to the extension property. I'm not sure how well KDoc would do referencing a generated property. Regarding the feasibility of creating such an annotation processor, I would say that it's very easy. The design and implementation would look very similar to this simple example annotation processor I created for a talk: https://github.com/anthonycr/Tagger

@msya do you think it would be a good idea to generate these using an annotation processor, or would you prefer to keep them in the source of the file? If you prefer to generate them, I can ticket the work.

anthonycr commented on models/src/main/java/com/vimeo/networking2/Cinema.kt:
```diff
@@ -42,9 +45,10 @@ data class Cinema(
      */
     @Json(name = "type")
-    val type: CinemaType? = null,
+    val type: String? = null,
```
used "programmedContentItem" instead due to recent renaming by @msya on another PR

anthonycr commented on models/src/main/java/com/vimeo/networking2/Album.kt:
```diff
@@ -58,9 +61,10 @@ data class Album(
      */
     @Json(name = "layout")
-    val layout: AlbumLayoutType? = null,
+    val layout: String? = null,
```
Naming patterns are now as follows for enums (it's slightly relaxed but basically follows this):
```kotlin
class ClassName {
  @Json(name = "enum_name")
  val enumName: String?
}

val ClassName.enumNameType: EnumNameType
```

For enums that represent the type of the class (e.g. `NotificationType`, `VideoBadgeType`), the naming pattern is strictly as follows:
```kotlin
class ClassName {
  @Json(name = "type")
  val rawType: String?
}

val ClassName.type: ClassNameType
```

anthonycr commented on models/src/main/java/com/vimeo/networking2/VideoContext.kt:
```diff
@@ -12,14 +15,22 @@ data class VideoContext(
      * The contextual resource type.
      */
-    @Json(name = "resource_key")
+    @Json(name = "resource_type")
```
i'm not sure

anthonycr commented:
@howardrigberg I created a ticket for documenting the enums as that work is quite involved, see here: https://vimean.atlassian.net/browse/VA-3704 I think I got every enum that you commented on.

I made several changes
- I fixed the naming of enums so that all enums now have `-Type` appended.
- I enabled the CI build on the `models` module.
- I renamed all the fields that have an `@Json(name = "type")` annotation to `rawType: String` and the extension to `type: Enum`. Other enums have normal naming scheme of `propertyName: String` and `propertyNameType: Enum`
- Created `SeasonType` enum as it was missing.
- Renamed `Tvod` classes to be consistently named... one was named `Tvoid`, some were named `TVod`.
- Added missing `@Json` annotations
- Added missing `@JsonClass` annotations
- Fixed formatting on a number of files.

anthonycr commented on models/src/main/java/com/vimeo/networking2/PurchaseInteraction.kt:
```diff
@@ -13,15 +15,23 @@ import com.vimeo.networking2.enums.PurchaseStatusType
 
     @Json(name = "options")
-    override val options: List<ApiOptionsType>? = null,
+    override val options: List<String>? = null,
```
This will hide the documentation it inherits from the `Interaction` interface, so i'm going to leave it as is.
